jupyterhub/user.py

```diff
@@ -336,6 +344,69 @@ def sync_groups(self, group_names):
             self.orm_user.groups = []
         self.db.commit()
 
+    def sync_roles(self, auth_roles):
+        """Synchronize roles with database"""
+        current_roles = {g.name for g in self.db.query(orm.Role).all()}
+
+        auth_roles_names = _get_role_names(auth_roles)
+
+        new_roles_names = set(auth_roles_names).difference(current_roles)
+        if new_roles_names:
+            self.log.info(f"Adding new roles {new_roles_names} to the database")
+        if auth_roles:
+            roles = self.db.query(orm.Role).all()
+            existing_roles = {g.name for g in roles}
+            for role_name in auth_roles_names:
+                if role_name not in existing_roles:
+                    # create roles that don't exist yet
+                    self.log.info(f"Creating new role {role_name}")
+
+                    role = _get_role_object(auth_roles, role_name)
+
+                    role_name = []
+                    if 'name' in role.keys():
+                        role_name = role['name']
+
+                    scopes = []
+                    if 'scopes' in role.keys():
+                        scopes = role['scopes']
+                        from .scopes import _check_scopes_exist
+
+                        try:
+                            _check_scopes_exist(scopes, who_for=f"role {role_name}")
+                        except:
+                            raise web.HTTPError(409, "One of the scopes does not exist")
+
+                    groups = []
+                    if 'groups' in role.keys():
+                        for group_name in role['groups']:
+                            group = orm.Group.find(self.db, group_name)
+                            if group is not None:
+                                groups.append(group)
+
+                    services = []
+                    if 'services' in role.keys():
+                        for service_name in role['services']:
+                            service = orm.Service.find(self.db, service_name)
+                            if service is not None:
+                                services.append(service)
+                    users = []
+                    if 'users' in role.keys():
+                        for user_name in role['users']:
+                            user = orm.User.find(self.db, user_name)
+                            if user is not None:
+                                users.append(user)
+                    orm_role = orm.Role(
+                        name=role_name,
```

Comment:
Creating the role directly does not validate the `role_name`. I suspect that `create_role` was not used to avoid triggering database commit too early, but I think it's possibly better to reconcile the two.